aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,163 @@
+#include "ATen/NativeFunctions.h"
+#include "ATen/ATen.h"
+#include <algorithm>
+#include <sstream>
+
+#include "ATen/cuda/AccumulateType.cuh"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+
+namespace at {
+namespace native {
+
+// Map the index of an element in tensor from 1D to nD
+__device__ __forceinline__
+void linear_index_to_indices(int64_t linear_index, int64_t* each_dim_len, int64_t total_dims, int64_t* indices) {
+  int64_t res = linear_index;
+  for (int64_t i = 0; i < total_dims; i++) {
+    int64_t indices_i = linear_index * total_dims + i;
+    indices[indices_i] = res / each_dim_len[i];
+    res = res % each_dim_len[i];
+  }
+}
+
+/*
+Map the index of an element in tensor from nD to 1D. A tensor is originally in nD shape,
+and 1D is the unfolded version of it (a vector).
+
+Example: given a 3D tensor
+[
+  [ [1,2], [3,4] ],
+  [ [5,6], [7,8] ]
+]
+
+Here element 3 has nD index = (0,1,0), and this corresponds to oneD index = 2
+*/
+__device__ __forceinline__
+int64_t indices_to_linear_index(int64_t* indices, int64_t total_dims, int64_t* each_dim_len, int64_t src_linear_index) {
+  int64_t dest_linear_index = 0;
+  for (int64_t i = 0; i < total_dims; i++) {
+    int64_t indices_i = src_linear_index * total_dims + i;
+    dest_linear_index += indices[indices_i] * each_dim_len[i];
+  }
+  return dest_linear_index;
+}
+
+template <typename scalar_t>
+__global__
+void flip_cuda_kernel(scalar_t* in_t, scalar_t* out_t, int64_t N, int64_t* dims, int64_t* indices,
+  int64_t flip_dims_size, int64_t* each_dim_len, int64_t* shape, int64_t total_dims) {
+
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+
+  linear_index_to_indices(linear_index, each_dim_len, total_dims, indices);
```

Comment:
We can get rid of the `nd indices array` thing altogether. How about just pass in the strides and work on `1d index`? We can solve the noncontiguous case as well.

Reply:
@SsnL Sorry I just started working on this. So the currently implementation will require a clone() from input tensor, which makes the returned tensor (flipped) contiguous, is this is the desired behavior? If so, maybe it is ok to convert the input tensor to contiguous before the flip? By 'solve the noncontiguous case', does it say to implement an in-place version so that both of input and returned tensors are noncontiguous?

Reply:
For this one, how should I approach it? any examples I should follow?

Reply:
Messaged you. LMK if you want additional help.